neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,398 @@
+            fno_skip=fno_skip,
+            conv_module=spectral_convolution,
+        )
+        # We have an option to make the last operator (MLP in regular
+        # Transformer block) permutation equivariant. i.e., applying the
+        # operator per variable or applying the operator on the whole channel
+        # (like regular FNO).
+        if permutation_eq:
+            self.mixer = FNOBlocks(
+                in_channels=self.mixer_token_codimension,
+                out_channels=self.mixer_token_codimension,
+                apply_skip=True,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(self.token_codimension)
+            self.norm2 = Normalizer(self.mixer_token_codimension)
+            self.mixer_out_normalizer = Normalizer(
+                self.mixer_token_codimension)
+
+        else:
+            self.mixer = FNOBlocks(
+                in_channels=codimension_size,
+                out_channels=codimension_size,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(codimension_size)
+            self.norm2 = Normalizer(codimension_size)
+            self.mixer_out_normalizer = Normalizer(codimension_size)
+
+    def compute_attention(self, xa, batch_size):
+        """
+        Compute the key-query-value variant of the attention matrix.
+
+        Assumes input ``xa`` has been normalized.
+        xa: torch.tensor. Has shape (b * t, d, h, w, ...) where,
+                        b is the batch_size,
+                        t is the number of tokens,
+                        d is the token codimension,
+                        and h, w, .. are the domain dimensions.
+        """
+
+        k = self.K(xa)
+        q = self.Q(xa)
+        v = self.V(xa)
+        assert k.size(
+            1) % self.n_heads == 0, "Number of channels in k, q, and v should be divisible by number of heads"
+
+        # reshaping '(b t) (n d) h w -> b n t (d h w ...)'
+
+        t = k.size(0) // batch_size  # Compute the number of tokens `t`
+        # Computer per head token codimension `d`
+        d = k.size(1) // self.n_heads
+
+        # reshaping '(b t) (n d) h w -> b n t d h w ...'
+        k = k.view(batch_size, t, self.n_heads, d, *k.shape[-self.num_dims:])
+        q = q.view(batch_size, t, self.n_heads, d, *q.shape[-self.num_dims:])
+        v = v.view(batch_size, t, self.n_heads, d, *v.shape[-self.num_dims:])
+
+        # permuating 'b n t d h w -> b t n d h w ...'
+        rearrangement = dict(
+            pattern=f'b t n d {" ".join(einsum_symbols[-i] for i in range(self.num_dims))} -> b n t d {" ".join(einsum_symbols[-i] for i in range(self.num_dims))}')
+        k = torch.einsum(rearrangement["pattern"], k)
+        q = torch.einsum(rearrangement["pattern"], q)
+        v = torch.einsum(rearrangement["pattern"], v)
```

Comment:
We don't need an einsum here, it's just a permutation of the 2nd and 3rd dimensions. You can do it with just `torch.transpose(k, 1, 2)`.

Reply:
Done